foubus.py

```diff
@@ -42,411 +53,552 @@
         resp.reason,
         resp.headers,
         len(resp.data),
     )
+
     fm = gtfs_realtime_pb2.FeedMessage.FromString(resp.data)
     with open("tripUpdates.textproto", "w") as f:
         f.write(str(fm))
+
+    feed_age = (datetime.datetime.now() - datetime.datetime.fromtimestamp(fm.header.timestamp)).total_seconds()
     logging.info(
         "TripUpdates header: %s (timestamp %s, age %d seconds)",
         text_format.MessageToString(fm.header, as_one_line=True),
         datetime.datetime.fromtimestamp(fm.header.timestamp),
-        (
-            datetime.datetime.now()
-            - datetime.datetime.fromtimestamp(fm.header.timestamp)
-        ).total_seconds(),
+        feed_age,
     )
     logging.info(
         "TripUpdates: %d entity, %d stop_time_update",
         len(fm.entity),
         sum(len(e.trip_update.stop_time_update) for e in fm.entity),
     )
 
-    tt["realtime"] = [False] * len(tt)
+    tt["realtime"] = False
+    updates_count = 0
 
-    updates = 0
     for entity in fm.entity:
-        assert entity.trip_update.trip.trip_id, str(entity)
-        if (tt["trip_id"] == entity.trip_update.trip.trip_id).any():
+        if not entity.trip_update.trip.trip_id:
+            continue
+
+        trip_id = entity.trip_update.trip.trip_id
+        trip_date = entity.trip_update.trip.start_date
+
+        if (tt["trip_id"] == trip_id).any():
             logging.info(
                 "trip_update for %s: %s: %d stop_time_update",
-                entity.trip_update.trip.trip_id,
+                trip_id,
                 text_format.MessageToString(entity.trip_update.trip, as_one_line=True),
                 len(entity.trip_update.stop_time_update),
             )
             last_stop_sequence = None
+
             for stu in entity.trip_update.stop_time_update:
-                # TODO: implement delay propagation
-                # https://gtfs.org/documentation/realtime/feed-entities/trip-updates/#:~:text=If%20one%20or%20more%20stops%20are%20missing
-                assert (
-                    last_stop_sequence is None
-                    or last_stop_sequence + 1 == stu.stop_sequence
-                ), text_format.MessageToString(stu, as_one_line=True)
-
-                if (
-                    stu.schedule_relationship
-                    != gtfs_realtime_pb2.TripUpdate.StopTimeUpdate.ScheduleRelationship.SCHEDULED
-                ):
+                # Validate consecutive stop sequences
+                if last_stop_sequence is not None and last_stop_sequence + 1 != stu.stop_sequence:
+                    logging.warning(
+                        "Non-consecutive stop sequence: %s",
+                        text_format.MessageToString(stu, as_one_line=True),
+                    )
+                last_stop_sequence = stu.stop_sequence
+
+                # Only apply updates for scheduled stops
+                if stu.schedule_relationship != stu.ScheduleRelationship.SCHEDULED:
```

Comment:
> stu.ScheduleRelationship.SCHEDULED

I don't think this is a best practice for protos. `stu` is a specific instance, and I think it's misleading to get a constant from an instance. I would prefer keeping the expression with _pb2.